lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,243 @@
+              if key == "type" && template_is_raw?(file_content) && !(@visited[str_url])
+                template = get_template_contents(str_url)
+                file_content = YAML.dump(template)
+              end
+
+              @files[str_url] = file_content
+              # replace the data value with the normalised absolute URL
+              Fog::Logger.debug("Replacing #{key} with #{str_url} in #{from_data}")
+              from_data[key] = str_url
+            end
+          end
+        end
+
+        private
+
+        # Retrive the content of a local or remote file.
+        #
+        # @param A local or remote uri.
+        #
+        # @raise ArgumentError if it's not a valid uri
+        #
+        # Protect open-uri from malign arguments like
+        #  - "|ls"
+        #  - multiline strings
+        def get_content(uri_or_filename)
+          remote_schemes = %w[http https ftp]
+          Fog::Logger.debug("Opening #{uri_or_filename}")
+
+          begin
+            # Validate URI to protect from open-uri attacks.
+            url = URI(uri_or_filename)
+
+            # Remote schemes must contain an host.
+            raise ArgumentError if url.host.nil? && remote_schemes.include?(url.scheme)
+
+            # Encode URI with spaces.
+            uri_or_filename = URI.encode(URI.decode(URI(uri_or_filename).to_s))
+          rescue URI::InvalidURIError
+            raise ArgumentError, "Not a valid URI: #{uri_or_filename}"
+          end
+
+          # TODO: A future revision may implement a retry.
+          # TODO: A future revision may limit download size.
+          content = ''
+          # open-uri doesn't open "file:///" uris.
+          uri_or_filename = uri_or_filename.sub(/^file:/, "")
+
+          open(uri_or_filename) { |f| content = f.read }
+          content
+        end
+
+        # Return true if the file is an heat template, false otherwise.
+        def template_is_raw?(content)
+          htv = content.index("heat_template_version:")
+          htv && htv < 5
+        end
+
+        # Return true if it's an URI, false otherwise.
+        def template_is_url?(path)
+          normalise_file_path_to_url(path)
+          true
+        rescue ArgumentError, URI::InvalidURIError
+          false
+        end
+
+        def file_outside_base_url?(base_url, str_url)
+          ret = base_url && str_url.start_with?("file://") && !str_url.start_with?(base_url)
+          if ret
+            Fog::Logger.warning("Trying to reference a file outside #{base_url}: #{str_url}")
+          end
+          ret
+        end
+
+        # Return true if I should I process this this file.
+        #
+        # @param [String] An heat template key
+        #
+        def ignore_if(key, value)
+          return true if key != 'get_file' && key != 'type'
+
+          return true unless value.kind_of?(String)
+
+          return true if key == 'type' &&
+                         !value.end_with?('.yaml', '.template')
```

Comment:
can it end with '.yml'? should it be case insensitive?

Reply:
@bzwei While it could end with yml:

- python implementation uses this more strict policy to avoid accidentally passing files to heat https://github.com/openstack/python-heatclient/blob/master/heatclient/common/template_utils.py#L112